discoverd/server/dns.go

```diff
@@ -6,32 +6,49 @@ import (
 	"net"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/miekg/dns"
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/vanillahsu/go_reuseport"
 	"github.com/flynn/flynn/discoverd/client"
 	"github.com/flynn/flynn/pkg/random"
 )
 
+type DNSStore interface {
+	Instances(service string) ([]*discoverd.Instance, error)
+	ServiceLeader(service string) (*discoverd.Instance, error)
+}
+
 type DNSServer struct {
 	UDPAddr   string
 	TCPAddr   string
 	Domain    string
 	Recursors []string
 
-	Store interface {
-		Instances(service string) ([]*discoverd.Instance, error)
-		ServiceLeader(service string) (*discoverd.Instance, error)
-	}
+	store    DNSStore
+	storeMtx sync.Mutex
```

Comment:
I looked into using `atomic.Value` but it seems that it can only store values of the same concrete type? We swap the store between the actual implementation and the proxy implementation which are different types. Is there a workaround that makes sense or given the limitations is the mutex just the more straightforward option?